discovery-provider/src/tasks/index_user_listening_history.py

```diff
@@ -0,0 +1,137 @@
+
+def _index_user_listening_history(session):
+    # get the last updated id that counted towards user_listening_history
+    # use as lower bound
+    prev_id_checkpoint = (session.query(IndexingCheckpoints.last_checkpoint)
+        .filter(IndexingCheckpoints.tablename == USER_LISTENING_HISTORY_TABLE_NAME)
+    ).scalar()
+
+    if not prev_id_checkpoint:
+        prev_id_checkpoint = 0
+
+    # get new plays since the last checkpoint
+    new_plays = (
+        session.query(Play.id, Play.user_id, Play.play_item_id, Play.created_at)
+        .filter(Play.id > prev_id_checkpoint)
+        .filter(Play.user_id != None)
+        .order_by(sa.asc(Play.id))
+        .limit(BATCH_SIZE)
+    ).all()
+
+    # no update exit early
+    if not new_plays:
+        return
+    new_checkpoint = new_plays[-1].id # get the highest play id
+
+    # get existing user listening history for users with new plays
+    users_with_new_plays = set([new_play.user_id for new_play in new_plays])
+    existing_user_listening_history = (
+        session.query(UserListeningHistory)
+        .filter(UserListeningHistory.user_id.in_(users_with_new_plays))
+        .all()
+    )
+    existing_users = set([user_history.user_id for user_history in existing_user_listening_history])
+
+    # reduce new plays
+    insert_user_listening_history_dict = DefaultDict(list)
+    update_user_listening_history_dict = DefaultDict(list)
+    for new_play in reversed(new_plays):
+        listen_obj ={"track_id": new_play.play_item_id, "timestamp": str(new_play.created_at)}
+        if new_play.user_id in existing_users:
+            update_user_listening_history_dict[new_play.user_id].append(listen_obj)
+        else:
+            insert_user_listening_history_dict[new_play.user_id].append(listen_obj)
+
+    # make updates to existing users
+    for i, user_history in enumerate(existing_user_listening_history):
+        deduped_history_dict = {}
+        for existing_play in reversed(user_history.listening_history):
+            deduped_history_dict[existing_play["track_id"]] = existing_play["timestamp"]
+        for new_play in reversed(update_user_listening_history_dict[user_history.user_id]):
+            deduped_history_dict[new_play["track_id"]] = new_play["timestamp"]
+        existing_user_listening_history[i].listening_history = sort_listening_history(deduped_history_dict)
+
+    # insert for new users
+    new_user_listening_history = []
+    for user, listening_history in insert_user_listening_history_dict.items():
+        deduped_history_dict = {}
+        for new_play in reversed(listening_history):
+            deduped_history_dict[new_play["track_id"]] = new_play["timestamp"]
+
+        new_user_listening_history.append(
+            UserListeningHistory(
+                user_id = user,
+                listening_history = sort_listening_history(deduped_history_dict)
+            )
+        )
+    session.add_all(new_user_listening_history)
```

Comment:
if i understand correctly whats happening here is
- for existing users the play is being appended directly to the UserListeningHistory record
- for new users the record is created and added to session

not sure of the perf effects of having so many db queries in a session, probably takes a while but most likely not blocking. but just thinking about other options, would it be possible to just do a single upsert query with all the record that have been created/modified?

Reply:
Yeah that's the behavior. I wanted to do a single query at first but couldn't figure it out. It might be possible with a really complex query and appending logic. I like how simple it is using SQLAlchemy but maybe there's a performance hit.

Since we're just doing insert and updates, it shouldn't block any queries on user_listening_history. Doing large reads on plays while first populating this table might be a concern?

Reply:
What if we deleted the existing rows and insert everything? That would be two queries in a tx but very simple

Reply:
Wouldn't that require looking at historical plays? Unless you first query for any affected users, update those histories. Then it becomes similar to this current method.

Reply:
Yeah I was thinking we would just update the existing record. Only benefit is you can do it in one query which might be slightly more efficient but otherwise it's basically the same